### PR TITLE
Add an action to setup environment

### PR DESCRIPTION
**Pull request by baizhenyu**
This only works on Debian based docker image.

### CODE REVIEW

File: setup-env/action.yml [hunk 1]
```diff
@@ -0,0 +1,21 @@
+name: 'Setup Environment'
+description: 'Setup dependencies for base container'
+
+runs:
+  using: "composite"
```
Comment (bemailloux) [inline setup-env/action.yml:5]: What does "composite" mean here?

File: setup-env/action.yml [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+
+        # Install git if not exist
+        if ! command -v git &> /dev/null; then
+          apt install git -y
+        fi
+
+        # Install awscli if not exist
+        if ! command -v aws &> /dev/null; then
+          apt install awscli -y
+        fi
```
Comment (bemailloux) [inline setup-env/action.yml:20]: Do we need to install other stuff like npm and node? Or is that already installed as part of the Docker image?

File: setup-env/action.yml (hunk 1 above)
Comment (baizhenyu) [reply setup-env/action.yml:5]: Github Actions have different types: https://docs.github.com/en/actions/creating-actions/about-actions#types-of-actions
Basically, it means run on the parent workflow's environment without creating a new environment.

File: setup-env/action.yml (hunk 2 above)
Comment (baizhenyu) [reply setup-env/action.yml:20]: It is been taken care of by other actions, e.g: https://github.com/Brightspace/bdp-batch/blob/US116893/githubAction/.github/workflows/ci.yml#L23

**Comment by SheranWiratungaD2L** [comment]
Is this needed? I think the default runner should have everything we need

**Comment by baizhenyu** [comment]
> Is this needed? I think the default runner should have everything we need

Running on the host directly might have some issues. I prefer to have a semi-isolated environment that does not rely on the host's environment.

File: setup-env/action.yml (hunk 2 above)
Comment (neverendingqs) [inline setup-env/action.yml:20]: Double-check: does `apt` have the latest versions of `git` and `awscli`? If not, would this be okay anyway?

Comment (baizhenyu) [reply setup-env/action.yml:20]: No, they are not the latest versions but they are supposed to be one of the stable versions:
https://packages.ubuntu.com/focal/awscli
https://packages.ubuntu.com/focal/git
Unless the CLI of them changes significantly. otherwise there should not be any problem.

**Comment by SheranWiratungaD2L** [comment]
> 
> 
> > Is this needed? I think the default runner should have everything we need
> 
> Running on the host directly might have some issues. I prefer to have a semi-isolated environment that does not rely on the host's environment.

I know the Maven build had to be in a container, are there other issues? I think checking out the repo and the AWS upload/pipeline run can be run on the host directly without issues, and then we can potentially have actions to run other steps in a container (i.e. the Maven build).

**Comment by baizhenyu** [comment]
> > > Is this needed? I think the default runner should have everything we need
> > 
> > 
> > Running on the host directly might have some issues. I prefer to have a semi-isolated environment that does not rely on the host's environment.
> 
> I know the Maven build had to be in a container, are there other issues? I think checking out the repo and the AWS upload/pipeline run can be run on the host directly without issues, and then we can potentially have actions to run other steps in a container (i.e. the Maven build).

Running on the host directly means any changes we made to the system would be persistent and will affect others' job. The Github Action runner only cleans the workspace after the job is completed. It means if we or others issue some commands like `apt upgrade xxx package`, or, modify some config under the home directory. These changes will not be reverted after the CI job is completed. To avoid the potential mysterious build failure in the future, I propose to run our entire job in a container. The containers created by other actions such as Maven Build (java-build) would be running as sibling containers of base container with the same workspace file system mounted.

**Comment by baizhenyu** [comment]
FYI, some discussion regarding the Github Action runner: https://d2l.slack.com/archives/C01248XBSCW/p1598301751044600
TL;DR we expect to get rid of the `Container` by the end of this week.

**Comment by baizhenyu** [comment]
Merging this PR for now. WIll deprecate this Action gradually once we can get rid of `Container`.
